conjur/credentials_from_file.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+
+"""
+CredentialsFromFile module
+
+This module holds the logic for writing user-specific credentials
+to a netrc file on the user's machine when they login
+"""
+
+# Builtins
+import logging
+import netrc
+import os
+import stat
+
+# Internals
+from conjur.constants import DEFAULT_NETRC_FILE
+
+class CredentialsFromFile:
+    """
+    CredentialsFromFile
+
+    This class holds logic when credentials are kept in the netrc
+    """
+    @classmethod
+    def save(cls, credential_data):
+        """
+        Method that writes user data to a netrc file
+        and updates permissions on the file
+        """
+        with open(DEFAULT_NETRC_FILE, 'w') as netrc_file:
+            # pylint: disable=line-too-long
+            netrc_file.write(f"machine {credential_data.machine}\nlogin {credential_data.login}\npassword {credential_data.password}\n")
+            netrc_file.close()
+
+        # Ensures that the netrc file is only available to owner
+        os.chmod(DEFAULT_NETRC_FILE, stat.S_IRWXU)
+
+    @classmethod
+    def remove(cls):
+        """
+        Method that removes netrc data from the file.
+        Triggered during a logout
+        """
+        if os.path.exists(DEFAULT_NETRC_FILE) and os.path.getsize(DEFAULT_NETRC_FILE) != 0:
+            open(DEFAULT_NETRC_FILE, "w").close()
+        else:
+            raise Exception("Please log in")
```

Comment:
???
This is not in this class responsibility to handle login/logout.
This method should remove the credentials form the file.
It's in the caller responsibility to handle failures.

Reply:
I moved it to logout_controller (the caller)

```
logout_controller
try:
    if os.path.exists(DEFAULT_NETRC_FILE) and os.path.getsize(DEFAULT_NETRC_FILE) != 0:
        self.logout_logic.remove()
        logging.debug("Logout successful")
        sys.stdout.write("Logged out from Conjur\n")
    else:
        raise Exception("Please log in")

# creds_from_file
def remove(cls):
    """
    Method that removes netrc data from the file.
    Triggered during a logout
    """
    open(DEFAULT_NETRC_FILE, "w").close()

```

Reply:
This error is super weird.
Better UX for logout is to fall for no-op in case user is not logged in.
Why telling the user to log in for log out to work?
Also, `Please log in` is not an error.
Error message should contain the info about the failure, in this case: `User is not logged in`.